### PR TITLE
Break up printer classes

### PR DESCRIPTION
**Pull request by Cellivar**
This is a first pass at breaking up the LP2844 class into logical modules. The goal is to make the various specific pieces able to be swapped out where appropriate:

* The USB communication layer should be minimized in the Printer class so it can eventually be swapped for WebSerial, Web Bluetooth, a network printer communication layer, etc.
* The EPL command set should be abstracted to a different layer so ZPL can be swapped in.
* The configuration loading system should be able to dynamically figure out EPL vs ZPL printers based on querying the config (and attempting different query commands until something works) then constructing the EPL/ZPL command set.

Adding a PrinterManager and rewriting the labels is left to a different PR.

### CODE REVIEW

File: src/LineBreakTransformer.ts [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
+        this.container += chunk;
+        const lines = this.container.split('\n');
+        this.container = lines.pop();
         lines.forEach((line) => controller.enqueue(line));
     }
 
-    flush(controller) {
-        controller.enqueue(this.#container);
+    flush(controller: TransformStreamDefaultController<string>) {
+        controller.enqueue(this.container);
```
Comment (Nican) [inline src/LineBreakTransformer.ts:12]: Add `this.container = ''` after flush?

File: src/UsbPrinterDeviceChannel.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        return this.inputStream;
+    }
+
+    constructor(device: USBDevice, enableConsoleDebug = false) {
+        super();
+
+        this.device = device;
+        this.EnableConsoleDebug = enableConsoleDebug;
+
+        this.ready = new Promise((resolve, reject) => {
```
Comment (Nican) [inline src/UsbPrinterDeviceChannel.ts:39]: This could also just be 
```this.ready = this.Connect();```

File: src/UsbPrinterDeviceChannel.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            console.debug('Sending print command buffer to printer via USB..');
+            console.debug(commandBuffer);
+            console.time('sendPrintBuffer');
+        }
+
+        try {
+            await this.device.transferOut(this.deviceOut.endpointNumber, commandBuffer);
+        } catch (e: unknown) {
+            const event = new CustomEvent<unknown>('usbPrinterDeviceOutError', e);
+            this.dispatchEvent(event);
```
Comment (Nican) [inline src/UsbPrinterDeviceChannel.ts:63]: I do not know the complete context around this, but why is this class an `EventTarget`?

**Review by Nican — COMMENTED** [review]
Some preliminary comments. Still learning your codebase.

File: src/Printers/Printer.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                    });
+                })
+                .catch(reject);
+        });
+    }
+
+    /**
+     * Refresh the printer information cache directly from the printer.
+     */
+    private async GetPrinterConfiguration() {
```
[the remark this answers is not in the document]
Comment (Cellivar) [reply src/Printers/Printer.ts:53]: Generally trying to [stick to this guide](https://google.github.io/styleguide/tsguide.html#identifiers) which is fairly C#-y that I'm getting it mostly correct by muscle memory. I haven't bothered setting up an ESLint rule for this yet and expected that to auto-format most stuff to be correct once I get around to it.

You're right though that that guide and your recommendation both call for functions and methods being `camelCase` unlike C#'s `PascalCase` for functions. By the time I'm ready to merge this it'll be normalized.

File: src/Printers/Printer.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        // Future TODO: Make this able to be dynamically changed out with
+        // WebSerial, WebBluetooth, etc. for other comm modes.
+        // Some kind of factory.
+        this.device = device;
+        this.printerChannel = new UsbPrinterDeviceChannel(
+            this.device,
+            this.PrinterCommunicationOptions.Debug
+        );
+
+        this.ready = new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (Cellivar) [reply src/Printers/Printer.ts:38]: Yeah this pattern was a first pass at [implementing this recommended pattern](https://stackoverflow.com/a/45070748/1309423) for constructor operations that need to defer ready status until later. I hadn't done a second look to see if there was a more modern pattern yet, thanks for the suggestion.

The goal with that linked pattern is to explicitly provide a promise interface that you can await on and I forgot that `async` methods fit that bill.

File: src/Printers/Printer.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+export class Printer {
+    // Printer communication handles
+    private device: USBDevice;
+    private printerChannel: IPrinterDeviceChannel;
+
+    private ready: Promise<boolean>;
+    /**
+     * A promise indicating this printer is ready to be used.
+     */
+    get Ready() {
```
[the remark this answers is not in the document]
Comment (Cellivar) [reply src/Printers/Printer.ts:17]: Readonly [doesn't make it to runtime](https://www.typescriptlang.org/docs/handbook/2/objects.html#readonly-properties) and won't [become a JS-style readonly property](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty). Having a getter-only property is a stronger guarantee of enforcement, so it's what I've been defaulting to.

There are other situations (such as changing the configuration) where the ready promise may need to be re-set to an in-progress operation which is [what that ready pattern](https://stackoverflow.com/a/45070748/1309423) allows for. I was going to explore using that, which would require setting outside the constructor.

File: src/UsbPrinterDeviceChannel.ts [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+            } else if (endpoint.direction == 'in') {
+                i = endpoint;
+            }
+        }
+
+        // For no apparent reason sometimes printers will omit to advertise the
+        // input endpoint. Sometimes they'll also omit the output endpoint. This
+        // attempts to handle those situations in a degraded mode.
+        if (!o) {
+            console.error('Failed to find an output for USB printer, cannot communicate!');
```
[the remark this answers is not in the document]
Comment (Cellivar) [reply src/UsbPrinterDeviceChannel.ts:91]: Definitely throw, hadn't gotten there yet. 

I hadn't thought of providing an expected communication mode. The big issue is that it's pretty unexpected when it happens. I've had printers that do it never, once in a blue moon, and one that works _fine_ but is _only_ unidirectional mode. It's a weird blend. I'll noodle more on this.

File: src/LineBreakTransformer.ts (hunk 1 above)
Comment (Cellivar) [reply src/LineBreakTransformer.ts:12]: Generally the examples I've found [for transformers](https://developer.mozilla.org/en-US/docs/Web/API/TransformStream/TransformStream) don't do that. Ones I have seen tend to terminate the controller, that is, close the written-to stream, which this doesn't do. I haven't caught any bugs related to this particular bit of code so far so I'm tempted to leave it as-is and keep this in mind as I start looking at better unit test coverage.

File: src/UsbPrinterDeviceChannel.ts (hunk 3 above)
Comment (Cellivar) [reply src/UsbPrinterDeviceChannel.ts:63]: Based on some googling this [appears to be the right way to implement things that raise events](https://hackwild.com/article/event-handling-techniques/) so other things can add listeners.

Still tweaking the design here a bit but I want to encapsulate the different type of connect/disconnect events that the WebUSB interface can end up generating. So the PrinterDeviceChannel encapsulates the WebUSB specific bits, bubbling them up to the Printer class, which translates them into external events that either the PrinterManager or a custom solution can hook into for displaying popups or something.

File: src/UsbPrinterDeviceChannel.ts (hunk 2 above)
Comment (Cellivar) [reply src/UsbPrinterDeviceChannel.ts:39]: That's the point of the Ready promise, generally. More conceptually I don't believe there's a compelling reason to have a channel that isn't open, that is, I couldn't come up with a strong reason to allow users of the library to separate creating a printer from connecting to a printer.

The vibe I was getting off some other libraries was that having a constructor and then expecting everyone to then call another method to set it up leads to issues with uninitialized objects being accessed or race conditions. By always firing off the async operation and then providing the hook point to await or chain onto was offered as a more robust solution for this sort of thing.

Exposing that promise externally also gives the hook point for receiving useful information, such as errors.

File: src/Printers/Printer.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Cellivar) [reply src/Printers/Printer.ts:53]: Lol turns out you have to just actually enable naming conventions as errors. Yeah this'll be normalized.

File: demo/ts-browser.js [hunk 8]
```diff
@@ -30,32 +30,63 @@ const workerFile = window.URL.createObjectURL(
   )
 );
 
+const load = sourceUrl => {
+  const xhr = XMLHttpRequest
```
Comment (JasonKaz) [inline demo/ts-browser.js:34]: Why not use native `fetch`? I haven't written an AJAX request like this in maybe 15 years hah.

**Comment by Cellivar** [comment]
Closing in favor of #12, which is easier to follow than this.

**Pull request closed without merging**